python_modules/dagster/dagster/_core/definitions/auto_materialize_policy.py

```diff
@@ -226,8 +229,19 @@ def without_rules(self, *rules_to_remove: "AutoMaterializeRule") -> "AutoMateria
 
     @public
     def with_rules(self, *rules_to_add: "AutoMaterializeRule") -> "AutoMaterializePolicy":
-        """Constructs a copy of this policy with the specified rules added."""
-        return self._replace(rules=self.rules.union(set(rules_to_add)))
+        """Constructs a copy of this policy with the specified rules added. If an instance of a
+        provided rule already exists in this policy, it will be replaced.
+        """
+        return self._replace(
```

Comment:
I think this function might be a bit more readable (and more performant, not that this really matters) like:

```python
new_rule_types = {type(rule) for rule in rules_to_add}
return self._replace(rules=new_rule_types.union(
    {rule for rule in self.rules if type(rule) not in new_rule_types}
))
```
```